[vtk[core]:x64-linux] build failure

**Host Environment**
- OS: Ubuntu 20.04.3 LTS on Windows 10 x86_64 (WSL)
- Compiler: g++ (Ubuntu 11.1.0-1ubuntu1~20.04) 11.1.0
- CMake version: cmake version 3.21.2

**To Reproduce**
Steps to reproduce the behavior:
`git clone https://github.com/microsoft/vcpkg`
`sudo ./vcpkg/bootstrap-vcpkg.sh`
`sudo vcpkg/vcpkg integrate install`
then hit build with vscode's cmake extension

**vcpkg.json**
```
{
    "name": "vtk-test",
    "version": "1.0",
    "description": "--",
    "dependencies": [
        {
            "name": "vtk",
            "features": [ ]
        }
    ]
}
```

**Expected Behavior**
Installs without error

**Failure logs** 
```
[build] -- Building x64-linux-dbg
[build] CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:155 (message):
[build]     Command failed: /home/alex/.local/lib/pypy3.6/site-packages/cmake/data/bin/cmake --build . --config Debug --target install -- -v -j33
[build]     Working Directory: /mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/vtk/x64-linux-dbg
[build]     See logs for more information:
[build]       /mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/vtk/install-x64-linux-dbg-out.log
[build] 
[build] Call Stack (most recent call first):
[build]   /mnt/d/Git/GitHub/vtk-test/build/vcpkg_installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_build.cmake:102 (vcpkg_execute_build_process)
[build]   /mnt/d/Git/GitHub/vtk-test/build/vcpkg_installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_install.cmake:41 (vcpkg_cmake_build)
[build]   ports/vtk/portfile.cmake:171 (vcpkg_cmake_install)
[build]   scripts/ports.cmake:142 (include)
[build] 
[build] 
[build] Error: Building package vtk:x64-linux failed with: BUILD_FAILED
[build] Please ensure you're using the latest portfiles with `git pull` and `./vcpkg update`, then
[build] submit an issue at https://github.com/Microsoft/vcpkg/issues including:
[build]   package: vtk[core]:x64-linux -> 9.0.3-pv5.9.1#3
[build]   vcpkg version: 2021-11-02-af04ebf6274fd6f7a941bff4662b3955c64f6f42
[build]   vcpkg-tool version: 603748b5d 2021-11-04 (33 minutes ago)
[build] 
[build] Additionally, attach any relevant sections from the log files above.
[build] -- Running vcpkg install - failed
[build] CMake Error at vcpkg/scripts/buildsystems/vcpkg.cmake:831 (message):
[build]   vcpkg install failed.  See logs for more information:
[build]   /mnt/d/Git/GitHub/vtk-test/build/vcpkg-manifest-install.log
[build] Call Stack (most recent call first):
[build]   build/CMakeFiles/3.21.2/CMakeSystem.cmake:6 (include)
[build]   CMakeLists.txt:2 (project)
[build] 
[build] 
[build] -- Configuring incomplete, errors occurred!
[build] See also "/mnt/d/Git/GitHub/vtk-test/build/CMakeFiles/CMakeOutput.log".
[build] See also "/mnt/d/Git/GitHub/vtk-test/build/CMakeFiles/CMakeError.log".
[build] make: *** [Makefile:604: cmake_check_build_system] Error 1
[build] Build finished with exit code 2
```
Full log is attached here:
[LOG.txt](https://github.com/microsoft/vcpkg/files/7477411/LOG.txt)


**Additional context**
/mnt/d/Git/GitHub/vtk-test/build/vcpkg-manifest-install.log
[vcpkg-manifest-install.log](https://github.com/microsoft/vcpkg/files/7477424/vcpkg-manifest-install.log)

/mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/vtk/install-x64-linux-dbg-out.log
[install-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/7477435/install-x64-linux-dbg-out.log)

/mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/vtk/config-x64-linux-dbg-out.log
[config-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/7477481/config-x64-linux-dbg-out.log)


/mnt/d/Git/GitHub/vtk-test/build/CMakeFiles/CMakeError.log
[CMakeError.log](https://github.com/microsoft/vcpkg/files/7477441/CMakeError.log)

/mnt/d/Git/GitHub/vtk-test/build/CMakeFiles/CMakeOutput.log
[CMakeOutput.log](https://github.com/microsoft/vcpkg/files/7477444/CMakeOutput.log)

./vcpkg list
```
No packages are installed. Did you mean `search`?
```

Hi @AlexLeungZ 

Thanks for posting this issue.

From install-x64-linux-dbg-out.log, we found this:

```
/mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/vtk/src/edd2e5233f-f32830fbf2.clean/Common/Core/vtkGenericDataArrayLookupHelper.h: In function ‘bool detail::isnan(T)’:
/mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/vtk/src/edd2e5233f-f32830fbf2.clean/Common/Core/vtkGenericDataArrayLookupHelper.h:52:26: error: ‘numeric_limits’ is not a member of ‘std’
   52 |   return has_NaN<T, std::numeric_limits<T>::has_quiet_NaN>::isnan(x);
```

Seems  like a bug in `vtk`, <limits.h> is required.

please see https://discourse.vtk.org/t/compilation-error-include-limits-required-in-several-files/6496.

https://gitlab.kitware.com/vtk/vtk/-/merge_requests/7611

```
/mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/vtk/src/edd2e5233f-f32830fbf2.clean/Common/Core/vtkGenericDataArrayLookupHelper.h: In function ‘bool detail::isnan(T)’:
/mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/vtk/src/edd2e5233f-f32830fbf2.clean/Common/Core/vtkGenericDataArrayLookupHelper.h:52:26: error: ‘numeric_limits’ is not a member of ‘std’
   52 |   return has_NaN<T, std::numeric_limits<T>::has_quiet_NaN>::isnan(x);
```

I successfully installed vtk with this method:

1. After the first failed build with VCPKG, you should have a file called `Kitware-VTK-xxx.tar.gz` in path `.../vcpkg/downloads`, in my case it is `.../vcpkg/downloads/Kitware-VTK-2959413ff190bc6e3ff40f5b6c1342edd2e5233f.tar.gz`
2. `cd` into that directory
3. Unzip the .tar.gz file, `unpigz xvf Kitware-VTK-xxx.tar.gz` then `tar xvf Kitware-VTK-xxx.tar`
4. Right now, you should have a directory called `VTK-xxx`
5. Edit the following files: `Common/Core/vtkGenericDataArray.h`, `Common/Core/vtkGenericDataArrayLookupHelper.h`, `Common/DataModel/vtkPiecewiseFunction.cxx`, `Filters/HyperTree/vtkHyperTreeGridThreshold.cxx`, `Rendering/Core/vtkColorTransferFunction.cxx`
6. Add these two header to those files` #include <stdexcept>`, `#include <limits>`
7. Save everything
8. zip the directory `VTK-xxx`, `tar cvf VTK-xxx.tar VTK-xxx`, then `pigz VTK-xxx.tar`
9. Rename the file as `Kitware-VTK-xxx.tar.gz`, which should be the same as what it is named in the downloaded path of vcpkg
10. Replace the old `Kitware-VTK-xxx.tar.gz` with your new one
11. Run build again, this time you should have a error something similar to this `expert hash: ...` `actual hash: ....`, and a message like this `Installing port from location: /mnt/d/Git/GitHub/vtk-test/vcpkg/buildtrees/versioning/versions/vtk/1b201a8e060b6c2688ccc2e96cae175f2d8af205`
12. Goto that directory, edit the `portfile.cmake` file, 
13. Change the `SHA512` value with the `actual hash` value
14. Now build the whole thing again, you should be good to go

Thx NancyLi for helping

--- a/ports/vtk/missing-limits.patch
+++ b/ports/vtk/missing-limits.patch
@@ -1,0 +1,48 @@
+diff --git a/Common/Core/vtkGenericDataArrayLookupHelper.h b/Common/Core/vtkGenericDataArrayLookupHelper.h
+index ab9d572..202aaa2 100644
+--- a/Common/Core/vtkGenericDataArrayLookupHelper.h
++++ b/Common/Core/vtkGenericDataArrayLookupHelper.h
+@@ -25,6 +25,7 @@
+ #include "vtkIdList.h"
+ #include <algorithm>
+ #include <cmath>
++#include <limits>
+ #include <unordered_map>
+ #include <vector>
+ 
+diff --git a/Common/DataModel/vtkPiecewiseFunction.cxx b/Common/DataModel/vtkPiecewiseFunction.cxx
+index 22eca0b..11086f1 100644
+--- a/Common/DataModel/vtkPiecewiseFunction.cxx
++++ b/Common/DataModel/vtkPiecewiseFunction.cxx
+@@ -22,6 +22,7 @@
+ #include <cassert>
+ #include <cmath>
+ #include <iterator>
++#include <limits>
+ #include <set>
+ #include <vector>
+ 
+diff --git a/Filters/HyperTree/vtkHyperTreeGridThreshold.cxx b/Filters/HyperTree/vtkHyperTreeGridThreshold.cxx
+index a16bb27..1052192 100644
+--- a/Filters/HyperTree/vtkHyperTreeGridThreshold.cxx
++++ b/Filters/HyperTree/vtkHyperTreeGridThreshold.cxx
+@@ -27,6 +27,7 @@
+ #include "vtkHyperTreeGridNonOrientedCursor.h"
+ 
+ #include <cmath>
++#include <limits>
+ 
+ vtkStandardNewMacro(vtkHyperTreeGridThreshold);
+ 
+diff --git a/Rendering/Core/vtkColorTransferFunction.cxx b/Rendering/Core/vtkColorTransferFunction.cxx
+index 55c046b..1be0291 100644
+--- a/Rendering/Core/vtkColorTransferFunction.cxx
++++ b/Rendering/Core/vtkColorTransferFunction.cxx
+@@ -21,6 +21,7 @@
+ #include <algorithm>
+ #include <cmath>
+ #include <iterator>
++#include <limits>
+ #include <set>
+ #include <vector>
+ 

--- a/ports/vtk/portfile.cmake
+++ b/ports/vtk/portfile.cmake
@@ -28,6 +28,7 @@ vcpkg_from_github(
         156fb524.patch
         d107698a.patch
         fix-gdal.patch
+        missing-limits.patch # This patch can be removed in next version. Since it has been merged to upstream via https://gitlab.kitware.com/vtk/vtk/-/merge_requests/7611
 )
 
 # =============================================================================
@@ -146,7 +147,7 @@ vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
 
 # We set all libraries to "system" and explicitly list the ones that should use embedded copies
 vcpkg_cmake_configure(
-    SOURCE_PATH ${SOURCE_PATH}
+    SOURCE_PATH "${SOURCE_PATH}"
     OPTIONS
         ${FEATURE_OPTIONS}
         ${VTK_FEATURE_OPTIONS}
@@ -268,9 +269,9 @@ if("paraview" IN_LIST FEATURES)
 endif()
 
 if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
-    if(EXISTS ${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/CMakeFiles/vtkpythonmodules/static_python) #python headers
+    if(EXISTS "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/CMakeFiles/vtkpythonmodules/static_python") #python headers
         file(GLOB_RECURSE STATIC_PYTHON_FILES "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/CMakeFiles/*/static_python/*.h")
-        file(INSTALL ${STATIC_PYTHON_FILES} DESTINATION ${CURRENT_PACKAGES_DIR}/include/vtk-${VTK_SHORT_VERSION})
+        file(INSTALL ${STATIC_PYTHON_FILES} DESTINATION "${CURRENT_PACKAGES_DIR}/include/vtk-${VTK_SHORT_VERSION}")
     endif()
 endif()
 

--- a/ports/vtk/vcpkg.json
+++ b/ports/vtk/vcpkg.json
@@ -1,7 +1,7 @@
 {
   "name": "vtk",
   "version-semver": "9.0.3-pv5.9.1",
-  "port-version": 3,
+  "port-version": 4,
   "description": "Software system for 3D computer graphics, image processing, and visualization",
   "homepage": "https://github.com/Kitware/VTK",
   "dependencies": [

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -7050,7 +7050,7 @@
     },
     "vtk": {
       "baseline": "9.0.3-pv5.9.1",
-      "port-version": 3
+      "port-version": 4
     },
     "vtk-dicom": {
       "baseline": "0.8.12",

--- a/versions/v-/vtk.json
+++ b/versions/v-/vtk.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "223022edbe63b3ecdb3e9204a4ad63ab61b36ece",
+      "version-semver": "9.0.3-pv5.9.1",
+      "port-version": 4
+    },
+    {
       "git-tree": "1b201a8e060b6c2688ccc2e96cae175f2d8af205",
       "version-semver": "9.0.3-pv5.9.1",
       "port-version": 3

